src/main/java/eu/dissco/core/digitalspecimenprocessor/service/AnnotationPublisherService.java

```diff
@@ -139,35 +144,68 @@ private List<AnnotationProcessingRequest> convertJsonPatchToAnnotations(
         annotationProcessingRequest.setOaHasBody(
-            buildBody(extractValueString(action), sourceSystemID));
+            buildBody(extractValueString(action.get(VALUE)), sourceSystemID));
+        annotations.add(annotationProcessingRequest);
       } else if (action.get("op").asText().equals("remove")) {
         annotationProcessingRequest.setOaMotivation(OaMotivation.ODS_DELETING);
         annotationProcessingRequest.setOaMotivatedBy(
             "Received delete information from Source System with id: " + sourceSystemID);
+        annotations.add(annotationProcessingRequest);
+      } else if (action.get("op").asText().equals("copy")) {
+        var digitalSpecimenJson = mapper.convertValue(
+            digitalSpecimenRecord.digitalSpecimenWrapper().attributes(), JsonNode.class);
+        var valueNode = digitalSpecimenJson.at(action.get("from").asText());
+        if (!valueNode.isMissingNode()) {
+          annotationProcessingRequest.setOaMotivation(OaMotivation.ODS_ADDING);
+          annotationProcessingRequest.setOaMotivatedBy(NEW_INFORMATION_MESSAGE + sourceSystemID);
+          annotationProcessingRequest.setOaHasBody(
+              buildBody(extractValueString(valueNode), sourceSystemID));
+          annotations.add(annotationProcessingRequest);
+        } else {
+          log.warn("Invalid copy operation in json patch: {} Ignoring this annotation", action);
+        }
+      } else if (action.get("op").asText().equals("move")) {
+        var digitalSpecimenJson = mapper.convertValue(
+            digitalSpecimenRecord.digitalSpecimenWrapper().attributes(), JsonNode.class);
+        var valueNode = digitalSpecimenJson.at(action.get("from").asText());
+        annotationProcessingRequest.setOaMotivation(OaMotivation.ODS_ADDING);
+        annotationProcessingRequest.setOaMotivatedBy(NEW_INFORMATION_MESSAGE+ sourceSystemID);
+        annotationProcessingRequest.setOaHasBody(
+            buildBody(extractValueString(valueNode), sourceSystemID));
+        var additionalDeleteAnnotation = new AnnotationProcessingRequest()
+            .withOaHasTarget(buildTarget(digitalSpecimenRecord,
+                buildSpecimenSelector(action.get("from").asText())))
+            .withDctermsCreated(Date.from(Instant.now()))
+            .withDctermsCreator(
+                new Agent().withType(Type.AS_APPLICATION).withId(sourceSystemID)
+                    .withSchemaName(sourceSystemName))
+            .withOaMotivation(OaMotivation.ODS_DELETING)
+            .withOaMotivatedBy(
+                "Received delete information from Source System with id: " + sourceSystemID);
+        annotations.add(additionalDeleteAnnotation);
+        annotations.add(annotationProcessingRequest);
       }
-      annotations.add(annotationProcessingRequest);
     }
     return annotations;
   }
 
   private String extractValueString(JsonNode action) throws JsonProcessingException {
```

Comment:
the input to this is: `digitalSpecimenJson.at(action.get("from").asText());`

you're checking if it is textual after you assign it to asText()

Reply:
No this isn't fed with the action.get("from") but with the valueNode (in the buildBody)
That node can be a string value but in the case of a class it can be a piece of Json, so that is when we use the writeValueAsString. If we do the writeValueAsString to a textNode we get double quotes ""sourceSystemName"".
```
        annotationProcessingRequest.setOaHasBody(
            buildBody(extractValueString(valueNode), sourceSystemID));
```

Reply:
see line 196: 
```java
var valueNode = digitalSpecimenJson.at(action.get(FROM).asText());
annotationProcessingRequest.setOaHasBody(
          buildBody(extractValueString(valueNode), sourceSystemID));
``` 
Is this intentional?

Reply:
I couldn't find this anymore, are you sure you got the latest version?

Reply:
i misread the parentheses, you're good